provider/datagen/src/lib.rs

```diff
@@ -256,30 +257,34 @@ pub fn keys_from_file<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<DataKey>>
 /// ```
 pub fn keys_from_bin<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<DataKey>> {
     let file = std::fs::read(path.as_ref())?;
-    let mut result = Vec::new();
-    let mut i = 0;
-    let mut last_start = None;
-    while i < file.len() {
-        if file[i..].starts_with(icu_provider::leading_tag!().as_bytes()) {
-            i += icu_provider::leading_tag!().len();
-            last_start = Some(i);
-        } else if file[i..].starts_with(icu_provider::trailing_tag!().as_bytes())
-            && last_start.is_some()
-        {
-            if let Some(key) = std::str::from_utf8(&file[last_start.unwrap()..i])
-                .ok()
-                .and_then(crate::key)
-            {
-                result.push(key);
-            }
-            i += icu_provider::trailing_tag!().len();
-            last_start = None;
-        } else {
-            i += 1;
-        }
+    let mut file = file.as_slice();
+
+    const LEADING_TAG: &[u8] = icu_provider::leading_tag!().as_bytes();
+    const TRAILING_TAG: &[u8] = icu_provider::trailing_tag!().as_bytes();
+
+    /// Given a needle like "abc" and a haystack like "123abc456", return ("123", "456")
+    fn split_at_tag<'a>(haystack: &'a [u8], needle: &[u8]) -> Option<(&'a [u8], &'a [u8])> {
+        memchr::memmem::find(haystack, needle)
```

Comment:
This repeatedly calls `memmem::find`, which the docs advise against: 
> Note that if you’re are searching for the same needle in many different small haystacks, it may be faster to initialize a [Finder](https://docs.rs/memchr/latest/memchr/memmem/struct.Finder.html) once, and reuse it for each search.

It probably makes sense to use `find_iter` for leading tags, as you're iterating over these anyway, and then have a `Finder` to find each leading tag's trailing tag.

Reply:
It occurred to me that a sufficiently zealous string optimizer might actually produce this as compiled output, which makes me wonder if these sorts of "partially overlapping" tags should be handled as such. This would be a larger divergence from the original algorithm, but I'm happy to add it anyway (it would result only in *more* tags being discovered than the original implementation).

To be specific, I'm proposing that this:

`"\nicu4x_key_tagTAG1\nicu4xkey_tagTAG2\n"`

should result in both of those tags being discovered, even though the second `\n` is being used simultaneously as trailing tag and as part of a leading tag.

The more I think about it the more I like this, so let me know your thoughts.

Reply:
> A find_iter implementation would discover no tags here (because the second \n would be perceived as a LEADING_TAG instead of a TRAILING_TAG).

I don't think that's true. `find_iter` returns a sequence of indices, then from each index we can search for the end tag, even if it overlaps with `find_iter`'s next output. I'm not sure if that's actually an improvement over using a `Finder` though.

I think your solution for partially overlapping tasks sounds good.

Reply:
For reference, my proposed updated version looks like this:

```rust
    let mut result: Vec<DataKey> = memmem::find_iter(file, LEADING_TAG)
        .map(|tag_position| tag_position + LEADING_TAG.len())
        .map(|key_start| &file[key_start..])
        .map(move |key_fragment| match trailing_tag.find(key_fragment) {
            Some(end) => &key_fragment[..end],
            None => key_fragment,
        })
        .filter_map(|key| std::str::from_utf8(key).ok())
        .filter_map(crate::key)
        .collect();
```

This would produce some weird results in degenerate overlap cases that are avoided by the original implementation and the current version in this PR, but if you agree that those are acceptable I'll update to this version.

Reply:
I think this is good, I didn't put much thought into the initial implementation and didn't actually think about the overlapping tag issue.

The only thing I'd change is to not return the whole rest of the string when no end tag is found:

```rust
        .filter_map(move |key_fragment| trailing_tag.find(key_fragment).map(|end| &key_fragment[..end])
```

Reply:
Done